File: swift/Tests/sdnxTests/SyntaxTests.swift
import Testing
@testable import sdnx

@Suite("Syntax error tests") struct SyntaxTests {
    @Test("no opening brace at top level") func noOpeningBrace() {
        let input = "age: 5"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Expected '{' but found 'a'" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("no closing brace at top level") func noClosingBrace() {
        let input = "{ age: 5 "
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Object not closed" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("no closing array brace") func noClosingArrayBrace() {
        let input = "{ foods: [\"ice cream\", \"strudel\" }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Array not closed" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("no field value") func noFieldValue() {
        let input = "{ foods, things: true }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Expected ':' but found ','" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("unsupported value type") func unsupportedValueType() {
        let input = "{ foods: things }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Unsupported value type 'things'" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("empty input") func emptyInput() {
        let input = ""
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Expected '{' but found 'EOF'" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("just whitespace") func justWhitespace() {
        let input = "   \n\t  "
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Expected '{' but found 'EOF'" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("field name starts with number") func fieldNameStartsWithNumber() {
        let input = "{ 1field: \"value\" }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Field must start with quote or alpha" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("field name with special chars") func fieldNameWithSpecialChars() {
        let input = "{ field-name: \"value\" }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Expected ':' but found '-'" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("unclosed string") func unclosedString() {
        let input = "{ name: \"Alice }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "String not closed" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("invalid escape sequence") func invalidEscapeSequence() {
        let input = "{ quote: \"Hel\\lo\" }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Invalid escape sequence '\\l'" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("number with decimal but no digits") func numberWithDecimalNoDigits() {
        let input = "{ value: 123. }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Unsupported value type '123.'" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("hex number with invalid chars") func hexNumberWithInvalidChars() {
        let input = "{ color: 0xGHIJKL }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Unsupported value type '0xGHIJKL'" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("invalid date format") func invalidDateFormat() {
        let input = "{ dob: 2025-13-01 }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Invalid date '2025-13-01'" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("boolean with wrong case") func booleanWrongCase() {
        let input = "{ active: True }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Unsupported value type 'True'" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("negative without digits") func negativeWithoutDigits() {
        let input = "{ value: - }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Unsupported value type '-'" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("scientific notation missing exponent") func scientificNotationMissingExponent() {
        let input = "{ value: 1.5e }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Unsupported value type '1.5e'" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("multiple colons in field") func multipleColonsInField() {
        let input = "{ name:: \"Alice\" }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Unsupported value type ':'" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("array missing separator") func arrayMissingSeparator() {
        let input = "{ items: [\"a\" \"b\"] }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Expected ',' but found '\"'" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("array with trailing comma not followed by item") func arrayTrailingCommaNoItem() {
        let input = "{ items: [\"a\",] }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Unsupported value type ''" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("nested object not closed") func nestedObjectNotClosed() {
        let input = "{ data: { nested: \"value\" }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Object not closed" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("nested array not closed") func nestedArrayNotClosed() {
        let input = "{ matrix: [[1, 2] }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Array not closed" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("object with missing colon after field name") func missingColonAfterFieldName() {
        let input = "{ name \"Alice\" }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Expected ':' but found '\"'" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("array with just opening brace") func arrayJustOpeningBrace() {
        let input = "{ items: [ }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Array not closed" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("array with missing opening brace") func arrayMissingOpeningBrace() {
        let input = "{ items: 1, 2, 3] }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Field must start with quote or alpha" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("field name with starting underscore") func fieldNameWithStartingUnderscore() {
        let input = "{ _private: \"value\" }"
        let result = parse(input)
        
        #expect((try? unwrapParseResult(result)) != nil, "Parse should succeed")
    }
    
    @Test("field name in quotes") func fieldNameInQuotes() {
        let input = "{ \"private-field\": \"hidden\" }"
        let result = parse(input)
        
        #expect((try? unwrapParseResult(result)) != nil, "Parse should succeed")
    }
    
    @Test("multiple commas in object") func multipleCommasInObject() {
        let input = "{ name: \"Alice\",, age: 30 }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Field must start with quote or alpha" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("comma at start of object") func commaAtStartOfObject() {
        let input = "{ , name: \"Alice\" }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Field must start with quote or alpha" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("invalid time format") func invalidTimeFormat() {
        let input = "{ time: 25:00 }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Invalid time '25:00'" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("invalid datetime format") func invalidDatetimeFormat() {
        let input = "{ created: 2025-01-15T14:90+02:00 }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Invalid date '2025-01-15T14:90+02:00'" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
    
    @Test("string with unescaped quote") func stringWithUnescapedQuote() {
        let input = "{ text: \"Hello \"World\"\" }"
        let result = parse(input)
        
        switch result {
        case .failure(let failure):
            #expect(failure.errors.contains(where: { $0.message == "Expected ':' but found '\"'" }))
        case .success:
            #expect(Bool(false), "Expected parse to fail")
        }
    }
}

File: dotnet/Sdnx.Tests/SyntaxTests.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Linq;
using Sdnx.Core;

namespace Sdnx.Tests;

[TestClass]
public class SyntaxTests
{
	[TestMethod]
	public void SyntaxErrors_NoOpeningBraceAtTopLevel()
	{
		const string input = "age: 5";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Expected '{' but found 'a'"));
	}

	[TestMethod]
	public void SyntaxErrors_NoClosingBraceAtTopLevel()
	{
		const string input = "{ age: 5 ";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Object not closed"));
	}

	[TestMethod]
	public void SyntaxErrors_NoClosingArrayBrace()
	{
		const string input = "{ foods: [\"ice cream\", \"strudel\" }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Array not closed"));
	}

	[TestMethod]
	public void SyntaxErrors_NoFieldValue()
	{
		const string input = "{ foods, things: true }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Expected ':' but found ','"));
	}

	[TestMethod]
	public void SyntaxErrors_UnsupportedValueType()
	{
		const string input = "{ foods: things }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Unsupported value type 'things'"));
	}

	[TestMethod]
	public void SyntaxErrors_EmptyInput()
	{
		const string input = "";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Expected '{' but found 'undefined'"));
	}

	[TestMethod]
	public void SyntaxErrors_JustWhitespace()
	{
		const string input = "   \n\t  ";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Expected '{' but found 'undefined'"));
	}

	[TestMethod]
	public void SyntaxErrors_FieldNameStartsWithNumber()
	{
		const string input = "{ 1field: \"value\" }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Field must start with quote or alpha"));
	}

	[TestMethod]
	public void SyntaxErrors_FieldNameWithSpecialChars()
	{
		const string input = "{ field-name: \"value\" }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Expected ':' but found '-'"));
	}

	[TestMethod]
	public void SyntaxErrors_UnclosedString()
	{
		const string input = "{ name: \"Alice }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "String not closed"));
	}

	[TestMethod]
	public void SyntaxErrors_InvalidEscapeSequence()
	{
		const string input = "{ quote: \"Hel\\lo\" }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Invalid escape sequence '\\l'"));
	}

	[TestMethod]
	public void SyntaxErrors_NumberWithDecimalButNoDigits()
	{
		const string input = "{ value: 123. }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Unsupported value type '123.'"));
	}

	[TestMethod]
	public void SyntaxErrors_HexNumberWithInvalidChars()
	{
		const string input = "{ color: 0xGHIJKL }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Unsupported value type '0xGHIJKL'"));
	}

	[TestMethod]
	public void SyntaxErrors_InvalidDateFormat()
	{
		const string input = "{ dob: 2025-13-01 }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Invalid date '2025-13-01'"));
	}

	[TestMethod]
	public void SyntaxErrors_BooleanWithWrongCase()
	{
		const string input = "{ active: True }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Unsupported value type 'True'"));
	}

	[TestMethod]
	public void SyntaxErrors_NegativeWithoutDigits()
	{
		const string input = "{ value: - }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Unsupported value type '-'"));
	}

	[TestMethod]
	public void SyntaxErrors_ScientificNotationMissingExponent()
	{
		const string input = "{ value: 1.5e }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Unsupported value type '1.5e'"));
	}

	[TestMethod]
	public void SyntaxErrors_MultipleColonsInField()
	{
		const string input = "{ name:: \"Alice\" }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Unsupported value type ':'"));
	}

	[TestMethod]
	public void SyntaxErrors_ArrayMissingSeparator()
	{
		const string input = "{ items: [\"a\" \"b\"] }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Expected ',' but found '\"'"));
	}

	[TestMethod]
	public void SyntaxErrors_ArrayWithTrailingCommaNotFollowedByItem()
	{
		const string input = "{ items: [\"a\",] }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Unsupported value type ''"));
	}

	[TestMethod]
	public void SyntaxErrors_NestedObjectNotClosed()
	{
		const string input = "{ data: { nested: \"value\" }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Object not closed"));
	}

	[TestMethod]
	public void SyntaxErrors_NestedArrayNotClosed()
	{
		const string input = "{ matrix: [[1, 2] }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Array not closed"));
	}

	[TestMethod]
	public void SyntaxErrors_ObjectWithMissingColonAfterFieldName()
	{
		const string input = "{ name \"Alice\" }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Expected ':' but found '\"'"));
	}

	[TestMethod]
	public void SyntaxErrors_ArrayWithJustOpeningBrace()
	{
		const string input = "{ items: [ }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Array not closed"));
	}

	[TestMethod]
	public void SyntaxErrors_ArrayWithMissingOpeningBrace()
	{
		const string input = "{ items: 1, 2, 3] }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Field must start with quote or alpha"));
	}

	[TestMethod]
	public void SyntaxErrors_FieldNameWithStartingUnderscore()
	{
		const string input = "{ _private: \"value\" }";
		var result = ParseData.Parse(input);
		Assert.IsTrue(result.Ok);
	}

	[TestMethod]
	public void SyntaxErrors_FieldNameInQuotes()
	{
		const string input = "{ \"private-field\": \"hidden\" }";
		var result = ParseData.Parse(input);
		Assert.IsTrue(result.Ok);
	}

	[TestMethod]
	public void SyntaxErrors_MultipleCommasInObject()
	{
		const string input = "{ name: \"Alice\",, age: 30 }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Field must start with quote or alpha"));
	}

	[TestMethod]
	public void SyntaxErrors_CommaAtStartOfObject()
	{
		const string input = "{ , name: \"Alice\" }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Field must start with quote or alpha"));
	}

	[TestMethod]
	public void SyntaxErrors_InvalidTimeFormat()
	{
		const string input = "{ time: 25:00 }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Invalid time '25:00'"));
	}

	[TestMethod]
	public void SyntaxErrors_InvalidDatetimeFormat()
	{
		const string input = "{ created: 2025-01-15T14:90+02:00 }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Invalid date '2025-01-15T14:90+02:00'"));
	}

	[TestMethod]
	public void SyntaxErrors_StringWithUnescapedQuote()
	{
		const string input = "{ text: \"Hello \"World\"\" }";
		var result = ParseData.Parse(input);
		Assert.IsFalse(result.Ok);
		Assert.IsNotNull(result.Errors.FirstOrDefault(e => e.Message == "Expected ':' but found '\"'"));
	}
}
